### PR TITLE
Add ROS on Azure Windows VM Template

### PR DESCRIPTION
**Pull request by seanyen**
### Best Practice Checklist
Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use a parameter for resource locations with the defaultValue set to resourceGroup().location
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values)
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment. 
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md

- [x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

### Changelog

* Add ROS on Azure Windows VM Template

### CODE REVIEW

**Comment by seanyen** [comment]
It seems the `https://raw.githubusercontent.com/Azure/...` reference is not automatically replaced by the pull request fork at CI time as it mentioned [here](https://github.com/Azure/azure-quickstart-templates/blob/c99d83208f566f75bcd3301e645f5982c4b0a7c3/.github/CONTRIBUTING.md#relinking).

So currently for this pull request, I am still using my owned hard-coded fork path [here](https://github.com/Azure/azure-quickstart-templates/pull/6505/commits/75f22792a2f9e0f933e79b4d718d73a5ebdb5cab) to make CI happy. Is it an known issue? Or how should I fix that.

File: ros-vm-windows/azuredeploy.json [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+                    "windowsConfiguration": {
+                        "enableAutomaticUpdates": true,
+                        "provisionVmAgent": true
+                    }
+                },
+                "licenseType": "Windows_Client",
+                "diagnosticsProfile": {
+                    "bootDiagnostics": {
+                        "enabled": true,
+                        "storageUri": "[reference(resourceId('Microsoft.Storage/storageAccounts', variables('diagnosticsStorageAccountName')), '2017-06-01').primaryEndpoints['blob']]"
```
Comment (bmoore-msft) [inline ros-vm-windows/azuredeploy.json]: update apiVersion

File: ros-vm-windows/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+            "apiVersion": "2019-03-01",
+            "location": "[variables('location')]",
+            "properties": {
+                "publisher": "Microsoft.Compute",
+                "type": "CustomScriptExtension",
+                "typeHandlerVersion": "1.8",
+                "autoUpgradeMinorVersion": true,
+                "settings": {
+                    "fileUris": [
+                        "https://raw.githubusercontent.com/seanyen/azure-quickstart-templates/ros-vm/ros-vm-windows/installROS.ps1"
```
Comment (bmoore-msft) [inline ros-vm-windows/azuredeploy.json]: See [BP guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#deployment-artifacts-nested-templates-scripts) for how to handle artifacts

File: ros-vm-windows/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+                    ],
+                    "commandToExecute": "[concat('powershell -ExecutionPolicy Unrestricted -file installROS.ps1')]"
+                }
+            },
+            "dependsOn": [
```
Comment (bmoore-msft) [inline ros-vm-windows/azuredeploy.json]: move this above properties

File: ros-vm-windows/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+        },
+        {
+            "condition": "[contains(parameters('virtualMachineSize'), 'Standard_NV')]",
+            "name": "nvidia-gpu-driver-windows",
+            "apiVersion": "2019-05-01",
+            "type": "Microsoft.Resources/deployments",
+            "properties": {
+                "mode": "Incremental",
+                "templateLink": {
+                    "uri": "https://gallery.azure.com/artifact/20161101/microsoft.nvidia-gpu-driver-windows-arm.1.1.0/Artifacts/MainTemplate.json"
```
Comment (bmoore-msft) [inline ros-vm-windows/azuredeploy.json]: this must be included with the sample

File: ros-vm-windows/azuredeploy.json [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+                "parameters": {
+                    "vmName": {
+                        "value": "[variables('virtualMachineName')]"
+                    },
+                    "location": {
+                        "value": "[variables('location')]"
+                    }
+                }
+            },
+            "dependsOn": [
```
Comment (bmoore-msft) [inline ros-vm-windows/azuredeploy.json]: move this above properties

File: ros-vm-windows/azuredeploy.json [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+        "networkInterfaceName": "[concat(parameters('projectName'), '-nic')]",
+        "networkSecurityGroupName": "[concat(parameters('projectName'), '-nsg')]",
+        "virtualNetworkName": "[concat(parameters('projectName'), '-vnet')]",
+        "publicIpAddressName": "[concat(parameters('projectName'), '-ip')]",
+        "virtualMachineName": "[concat(parameters('projectName'), '-vm')]",
+        "diagnosticsStorageAccountName": "[concat(uniqueString(resourceGroup().id),'diag')]",
+        "nsgId": "[resourceId(resourceGroup().name, 'Microsoft.Network/networkSecurityGroups', variables('networkSecurityGroupName'))]",
+        "vnetId": "[resourceId(resourceGroup().name,'Microsoft.Network/virtualNetworks', variables('virtualNetworkName'))]",
+        "subnetName": "default",
+        "subnetRef": "[concat(variables('vnetId'), '/subnets/', variables('subnetName'))]",
```
Comment (bmoore-msft) [inline ros-vm-windows/azuredeploy.json]: remove concat and just use resourceId

[resourceId(resourceGroup().name,'Microsoft.Network/virtualNetworks/subnets', variables('virtualNetworkName'), variables('subnetName'))]

File: ros-vm-windows/azuredeploy.json [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+    },
+    "variables": {
+        "location": "[parameters('location')]",
+        "networkInterfaceName": "[concat(parameters('projectName'), '-nic')]",
+        "networkSecurityGroupName": "[concat(parameters('projectName'), '-nsg')]",
+        "virtualNetworkName": "[concat(parameters('projectName'), '-vnet')]",
+        "publicIpAddressName": "[concat(parameters('projectName'), '-ip')]",
+        "virtualMachineName": "[concat(parameters('projectName'), '-vm')]",
+        "diagnosticsStorageAccountName": "[concat(uniqueString(resourceGroup().id),'diag')]",
+        "nsgId": "[resourceId(resourceGroup().name, 'Microsoft.Network/networkSecurityGroups', variables('networkSecurityGroupName'))]",
```
Comment (bmoore-msft) [inline ros-vm-windows/azuredeploy.json]: remove resourceGroup().name

File: ros-vm-windows/azuredeploy.json [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+        "vnetId": "[resourceId(resourceGroup().name,'Microsoft.Network/virtualNetworks', variables('virtualNetworkName'))]",
```
Comment (bmoore-msft) [inline ros-vm-windows/azuredeploy.json]: remove this var after you update subnetRef below

File: ros-vm-windows/azuredeploy.json [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+        "publicIpAddressSku": "Basic",
+        "osDiskType": "StandardSSD_LRS"
+    },
+    "resources": [
+        {
+            "name": "[variables('networkInterfaceName')]",
+            "type": "Microsoft.Network/networkInterfaces",
+            "apiVersion": "2018-10-01",
+            "location": "[variables('location')]",
+            "dependsOn": [
```
Comment (bmoore-msft) [inline ros-vm-windows/azuredeploy.json]: remove concat from all dependsOn - use resourceId or just the resource name (e.g. [variables('virtualNetworkName')] )

File: ros-vm-windows/azuredeploy.json [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+                "ipConfigurations": [
+                    {
+                        "name": "ipconfig1",
+                        "properties": {
+                            "subnet": {
+                                "id": "[variables('subnetRef')]"
+                            },
+                            "privateIPAllocationMethod": "Dynamic",
+                            "publicIpAddress": {
+                                "id": "[resourceId(resourceGroup().name, 'Microsoft.Network/publicIpAddresses', variables('publicIpAddressName'))]"
```
Comment (bmoore-msft) [inline ros-vm-windows/azuredeploy.json]: remove resourceGroup().name

File: ros-vm-windows/azuredeploy.parameters.json [hunk 11]
```diff
@@ -0,0 +1,18 @@
+{
+    "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentParameters.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters": {
+        "projectName": {
+            "value": "GEN-UNIQUE-8"
+        },
+        "virtualMachineSize": {
```
Comment (bmoore-msft) [inline ros-vm-windows/azuredeploy.parameters.json]: remove this and set as the defaultValue in the template

File: ros-vm-windows/installROS.ps1 [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+
+$response = Invoke-RestMethod $latestBuildUrl -Method Get -ContentType 'application/json'
+$buildId = $response.value[0].id
+
+$artifactUrl = "https://ros-win.visualstudio.com/bed058dd-46da-4029-bb85-25eae7674d09/_apis/build/builds/$buildId/artifacts?artifactName=sfx-installer&api-version=5.2-preview.5&%24format=zip"
+
+New-Item -ItemType directory -Path $outputPath
+
+$download = New-Object net.webclient
+$download.Downloadfile($artifactUrl, $downloadPath)
```
Comment (bmoore-msft) [inline ros-vm-windows/installROS.ps1]: this is pretty big, let's add some error handling and retries around it

**Comment by seanyen** [comment]
@bmoore-msft Thank for your feedback! I addressed them and it seemed that now the CI result looks cleaner. Please take a look. Thanks!

**Comment by seanyen** [comment]
@bmoore-msft Thanks for the merge. One quick question, how long does it usually take to show up a new template in this portal? https://azure.microsoft.com/en-us/resources/templates/

**Pull request merged**